File: src/main/java/com/ject/studytrip/global/security/JwtAuthenticationFilter.java
package com.ject.studytrip.global.security;

import com.ject.studytrip.auth.application.service.TokenService;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.util.StringUtils;
import org.springframework.web.filter.OncePerRequestFilter;

@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {
    private final TokenService tokenService;

    @Override
    protected void doFilterInternal(
            @NonNull HttpServletRequest request,
            @NonNull HttpServletResponse response,
            @NonNull FilterChain filterChain)
            throws ServletException, IOException {
        String authorizationHeader = request.getHeader(HttpHeaders.AUTHORIZATION);
        String accessToken = extractBearerToken(authorizationHeader);

        if (StringUtils.hasText(accessToken)) {
            tokenService.validateActiveAccessToken(accessToken);
            tokenService.setAuthenticationByAccessToken(accessToken);
        }

        filterChain.doFilter(request, response);
    }

    private String extractBearerToken(String header) {
        return (StringUtils.hasText(header) && header.startsWith("Bearer "))
                ? header.substring(7)
                : null;
    }
}

File: src/main/kotlin/com/ject/studytrip/global/security/filter/JwtAuthenticationFilter.kt
package com.ject.studytrip.global.security.filter

import com.ject.studytrip.auth.application.service.TokenService
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpHeaders
import org.springframework.web.filter.OncePerRequestFilter

class JwtAuthenticationFilter(
    private val tokenService: TokenService,
) : OncePerRequestFilter() {
    companion object {
        private const val BEARER_PREFIX = "Bearer "
    }

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        extractBearerToken(request)
            ?.let { accessToken ->
                tokenService.validateActiveAccessToken(accessToken)
                tokenService.setAuthenticationByAccessToken(accessToken)
            }

        filterChain.doFilter(request, response)
    }

    private fun extractBearerToken(request: HttpServletRequest): String? {
        val header = request.getHeader(HttpHeaders.AUTHORIZATION)

        if (header.isNullOrBlank()) return null
        if (!header.startsWith(BEARER_PREFIX)) return null

        return header
            .removePrefix(BEARER_PREFIX)
            .trim()
            .takeIf { it.isNotBlank() }
    }
}
